### PR TITLE
[mail-composer][android] fix composeAsync not resolving after send/ discard

### LINKED ISSUE #20509: MailComposer.composeAsync doesn't return result on Android
### Summary

This is the same issue as https://github.com/expo/expo/issues/16690 and https://github.com/expo/expo/issues/16696, but both of those issues have been closed without fixing the issue. 

### What platform(s) does this occur on?

Android

### Environment

  expo-env-info 1.0.5 environment info:
    System:
      OS: macOS 12.6
      Shell: 5.8.1 - /bin/zsh
    Binaries:
      Node: 16.10.0 - /usr/local/bin/node
      Yarn: 1.22.17 - /usr/local/bin/yarn
      npm: 7.24.0 - /usr/local/bin/npm
    Managers:
      CocoaPods: 1.11.2 - /usr/local/bin/pod
    SDKs:
      iOS SDK:
        Platforms: DriverKit 22.2, iOS 16.2, macOS 13.1, tvOS 16.1, watchOS 9.1
    IDEs:
      Android Studio: 2020.3 AI-203.7717.56.2031.7935034
      Xcode: 14.2/14C18 - /usr/bin/xcodebuild
    npmPackages:
      @expo/webpack-config: ^0.17.2 => 0.17.3 
      babel-preset-expo: ~9.2.1 => 9.2.2 
      expo: ~47.0.8 => 47.0.8 
      react: 18.1.0 => 18.1.0 
      react-dom: 18.1.0 => 18.1.0 
      react-native: 0.70.5 => 0.70.5 
      react-native-web: ~0.18.7 => 0.18.9 
    npmGlobalPackages:
      eas-cli: 0.55.1
      expo-cli: 6.0.8
    Expo Workflow: managed

### Minimal reproducible example

I made a Snack: https://snack.expo.dev/@reachvote/mailcomposer. When I run this in Expo Go on my android device, I expect to get an alert with `result.status` after sending or not sending the email. This does work on iOS devices, but not on android.

I am aware that (according to documentation) Android always sets the status as if the email were sent, but the value of status is not the issue, the fact that no `result` is returned is.

### PR DESCRIPTION
**Pull request by keith-kurak**
# Why

Fixes https://github.com/expo/expo/issues/20509

# How

TL;DR - this fix restores the documented behavior, albeit with a visual bug that seems to be peculiar to Gmail.

I first tried to compare the send/ resolve flow to (what I thought) was the closest analog, `expo-sms`. The `onHostResume` lifecycle listener event is trigged in the SMS package, but not mail composer. Nothing along these lines worked (e.g., matching the Intent flags).

It turns out that `expo-sharing` is perhaps a more appropriate comparison, because they both use `Intent.createChooser` to create a share sheet (the mail composer's sheet mainly filters on email apps and can pass additional email-related keys). It seems like `onHostResume` never responds to returning from a share sheet. Sharing uses the `ActivityEventListener` instead.

Anyway, this works, with one visual quirk that seems to be peculiar to the implementation of the Gmail client's new mail activity. Pressing back does not discard the email/ return to the app. It takes you back to your inbox. Once you do this, your only choice is to use the task switcher/ launcher to leave the email, and, in the task switcher, **the image of the app doesn't update, continuing to show a picture of the email that you were composing**. 

Other apps demonstrate the same issue with sharing to Gmail (e.g., Twitter), though others don't (e.g., Facebook). Seems possible that this is related to task/ activity settings outside the control of this package, e.g., `android:launchMode`.

Comparing to external packages that fulfill a similar purpose, I found some (e.g., `react-native-mail`) that resolve the promise as soon as the share sheet opens, which is an option, but would break the documented behavior and not match iOS.

# Test Plan

Scenarios:
- [x] single email client (bypasses chooser)
- [x] multiple email clients (shows chooser)
- [x] send email (returns to app, promise resolves)
- [x] discard email (returns to app, promise resolves)
- [x] share email, open task switcher, return to app (promise resolves, but task switcher freezes on email client)
- [x] test with latest Expo Go unversioned as of 2/1/2023 

[See video](https://photos.app.goo.gl/a78ckUT68TLTfAne9) (including visual bug at the end)

# Checklist

- [x] Documentation is up to date to reflect these changes (eg: https://docs.expo.dev and README.md).
- [ ] Conforms with the [Documentation Writing Style Guide](https://github.com/expo/expo/blob/main/guides/Expo%20Documentation%20Writing%20Style%20Guide.md)
- [x] This diff will work correctly for `expo prebuild` & EAS Build (eg: updated a module plugin).

### CODE REVIEW

File: packages/expo-mail-composer/android/src/main/java/expo/modules/mailcomposer/MailComposerModule.kt [hunk 1]
```diff
@@ -65,25 +73,26 @@ class MailComposerModule(
...
 
-  override fun onHostPause() = Unit
-
-  override fun onHostDestroy() {
-    // do nothing
+  companion object {
+    private const val REQUEST_CODE = 8675
   }
-}
+}
```
Comment (lukmccall) [inline packages/expo-mail-composer/android/src/main/java/expo/modules/mailcomposer/MailComposerModule.kt:98]:
```suggestion
}

```

**Review by lukmccall — APPROVED** [review]
LGTM ✅
